internal/admission-webhook-server/admission-webhook-server.go

```diff
@@ -0,0 +1,146 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package admissionwebhookserver contains code for admission webhook.
+package admissionwebhookserver
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// ServeMutateServiceAccount is a service to mutate default service account.
+func ServeMutateServiceAccount(w http.ResponseWriter, r *http.Request) {
+	serve(w, r, mutateServiceAccount)
+}
+
+type patchOperation struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+type admitFunc func(v1.AdmissionReview) *v1.AdmissionResponse
+
+func toAdmissionResponse(err error) *v1.AdmissionResponse {
+	return &v1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+		},
+	}
+}
+
+func validateRequest(r *http.Request) ([]byte, error) {
+	var body []byte
+
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	if len(body) == 0 {
+		return nil, errors.New("empty body")
+	}
+
+	// Verify the content type is accurate.
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		return nil, errors.New("Content-Type=" + contentType + ", expect application/json")
+	}
+
+	return body, nil
+}
+
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	body, err := validateRequest(r)
+	if err != nil {
+		glog.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	// The AdmissionReview that was sent to the webhook.
+	requestedAdmissionReview := v1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned.
+	responseAdmissionReview := v1.AdmissionReview{}
+
+	deserializer := scheme.Codecs.UniversalDeserializer()
+
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		glog.Error(err)
+		responseAdmissionReview.Response = toAdmissionResponse(err)
+	} else {
+		responseAdmissionReview.Response = admit(requestedAdmissionReview)
+	}
+
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	responseAdmissionReview.APIVersion = "admission.k8s.io/v1"
+	responseAdmissionReview.Kind = "AdmissionReview"
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		glog.Error(err)
```

Comment:
It would be good to wrap those errors with some formatting to know what failed when reading the logs.